src/ffi.rs

```diff
@@ -1,7 +1,38 @@
 use mpz::*;
+use std::{mem, slice, str};
+use libc::{c_char, strlen};
+
+type AllocFunc = extern "C" fn(usize) -> *mut c_char;
+type ReallocFunc = extern "C" fn(*mut c_char, usize, usize) -> *mut c_char;
+type FreeFunc = extern "C" fn(*mut c_char, usize);
 
 #[link(name = "gmp")]
 extern "C" {
     pub fn __gmpz_fdiv_q(q: mpz_ptr, n: mpz_srcptr, d: mpz_srcptr);
     pub fn __gmpz_cdiv_q(q: mpz_ptr, n: mpz_srcptr, d: mpz_srcptr);
+    pub fn __gmp_get_memory_functions(alloc: *mut AllocFunc, relloc: *mut ReallocFunc, free: *mut FreeFunc);
+}
+
+pub struct GString(*mut u8, usize);
+
+impl GString {
+    pub unsafe fn from_raw(raw: *mut c_char) -> GString {
+        GString(mem::transmute(raw), strlen(raw) as usize + 1)
+    }
+
+    pub fn to_str(&self) -> Result<&str, str::Utf8Error> {
+        let bytes: &[u8] = unsafe { slice::from_raw_parts(self.0, self.1) };
+        str::from_utf8(&bytes[..bytes.len() - 1])
```

Comment:
no matter what, the length needs to be known in order to call the free func, so it's either strlen or CStr::to_bytes().len(). i think strlen communicates more clearly how the length param to free_func is computed, whereas the alternate method relies on knowing that CStr implicitly calls strlen behind the scenes.